lib/shoppingcart.rb

```diff
@@ -0,0 +1,59 @@
+require './lib/shoppingcart'
+
+class ShoppingCart
+  attr_reader :name,
+              :capacity,
+              :products
+
+
+  def initialize(name, capacity)
+    @name = name
+    @capacity = capacity.to_i
+    @products = []
+    @details = nil
+  end
+
+  def add_product(product)
+    @products << product
+  end
+
+  def details
+    @details = {
+      name: @name,
+      capacity: @capacity
+    }
+  end
+
+  def is_full?
+    (@products.sum do |product|
+      product.quantity
+    end) > @capacity
+  end
+
+  def products_by_category(category)
+    @products.find_all do |product|
+      product.category == category
+    end
+  end
+
+  def percentage_occupied
+    ((@products.sum do |product|
+      product.quantity
+    end).to_f / @capacity) * 100
+  end
```

Comment:
I think you may have accidentally skipped over creating the `total_number_of_products` method. You probably noticed that you're repeating the code on lines 18-30 and 40-42.. that code could go right into that method.

Reply:
also you may need to round the return value of `percentage_occupied`